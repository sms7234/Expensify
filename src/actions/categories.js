import database from '../firebase/firebase';

export const setCategories = (categories) => ({
  type:"SET_CATEGORIES",
  categories
});

export const startSetCategories = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`)
      .once('value')
      .then((snapshot) => {
        const categories = [];
        snapshot.forEach((child) => {
          categories.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setCategories(categories));
      });
  };
};

export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) =>{
    const {
      category= '',
      description= ''
    } = categoryData;
    const categoryItem = {category, description};
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`).push(categoryItem).then((ref)=>{
      dispatch(addCategory({
        id: ref.key,
        ...categoryItem
      }));
    });
  };
};

export const editCategory = (id, updates) => ({
  type: 'EDIT_CATEGORY',
  id,
  updates
});

export const startEditCategory = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).update(updates).then(()=>{
      dispatch(editCategory(id,updates));
    });
  };
};

export const removeCategory = (id) => ({
  type: 'REMOVE_CATEGORY',
  id
});

export const startRemoveCategory = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).remove().then(()=>{
      dispatch(removeCategory(id));
    });
  };
};
