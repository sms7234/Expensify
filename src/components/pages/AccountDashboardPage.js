import React from 'react';
import AccountList from '../lists/AccountList';
import AccountSummary from '../independents/AccountSummary';
import AccountListFilters from '../filters/AccountListFilters';

const AccountDashboardPage = () => (
  <div>
    <AccountSummary />
    <AccountListFilters />
    <AccountList />
  </div>
)

export default AccountDashboardPage;
