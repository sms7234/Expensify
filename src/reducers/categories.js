export default (state = [], action) => {
  switch(action.type){
    case 'ADD_CATEGORY':
      return state.concat(action.category);
    case 'SET_CATEGORIES':
      return action.categories;
    default:
      return state;
  }
};
