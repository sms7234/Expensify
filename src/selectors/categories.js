export default (categories, {category}) => {
  return categories.filter((item) => {
    const textMatch = item.category.toLowerCase().includes(category.toLowerCase());
    return textMatch;
  }).sort((a,b)=>{
    if(a.category < b.category){
      return -1;
    } else {
      return 1;
    }
  });
};
