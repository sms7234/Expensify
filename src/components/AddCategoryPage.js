import React from 'react';
import CategoryForm from './CategoryForm';
import {connect} from 'react-redux';
import {startAddCategory} from '../actions/categories';

export class AddCategoryPage extends React.Component{
  onSubmit=(category) => {
    this.props.startAddCategory(category);
    this.props.history.push('/categories');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Category </h1>
          </div>
        </div>
        <div className="content-container">
          <CategoryForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddCategory:(category) => dispatch(startAddCategory(category))
});

export default connect(undefined, mapDispatchToProps)(AddCategoryPage);
