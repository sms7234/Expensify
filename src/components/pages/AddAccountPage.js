import React from 'react';
import AccountForm from '../forms/AccountForm';
import {connect} from 'react-redux';
import {startAddAccount} from '../../actions/accounts';

export class AddAccountPage extends React.Component{
  onSubmit=(account) => {
    this.props.startAddAccount(account);
    this.props.history.push('/accounts');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Account </h1>
          </div>
        </div>
        <div className="content-container">
          <AccountForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddAccount:(account) => dispatch(startAddAccount(account))
});

export default connect(undefined, mapDispatchToProps)(AddAccountPage);
