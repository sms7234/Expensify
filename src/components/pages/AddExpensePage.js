import React from 'react';
import ExpenseForm from '../forms/ExpenseForm';
import { connect} from 'react-redux';
import {startAddExpense} from '../../actions/expenses';

export class AddExpensePage extends React.Component{
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/expenses');
  };
  render(){
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit = {this.onSubmit}
            categoryList = {this.props.categoryList}
            accountList = {this.props.accountList}
            tagList = {this.props.tagList}
          />
        </div>
      </div>
    );
  }
};

const mapStateToProps=(state) => {
  return {
    accountList: state.accounts.map((item)=>{return({label: item.account, value: item.account})}),
    categoryList: state.categories.map((item)=>{return({label: item.category, value: item.category})}),
    tagList: state.tags.map((item)=>{return({label: item.tag, value: item.tag})})
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);
