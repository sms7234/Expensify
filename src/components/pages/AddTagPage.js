import React from 'react';
import TagForm from '../forms/TagForm';
import {connect} from 'react-redux';
import {startAddTag} from '../../actions/tags';

export class AddTagPage extends React.Component{
  onSubmit=(tag) => {
    this.props.startAddTag(tag);
    this.props.history.push('/tags');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Tag </h1>
          </div>
        </div>
        <div className="content-container">
          <TagForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddTag:(tag) => dispatch(startAddTag(tag))
});

export default connect(undefined, mapDispatchToProps)(AddTagPage);
