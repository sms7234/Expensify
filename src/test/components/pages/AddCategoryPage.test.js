import React from 'react';
import {shallow} from 'enzyme';
import {AddCategoryPage} from '../../../components/pages/AddCategoryPage';
import categories from '../../fixtures/categories';

let startAddCategory, history, wrapper;

beforeEach(()=> {
  startAddCategory = jest.fn();
  history = {push: jest.fn()};
  wrapper = shallow(<AddCategoryPage startAddCategory={startAddCategory} history={history} />);
});

test('should render AddCategoryPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit',() => {
  wrapper.find('CategoryForm').prop('onSubmit')(categories[0]);
  expect(history.push).toHaveBeenLastCalledWith('/categories');
  expect(startAddCategory).toHaveBeenLastCalledWith(categories[0]);
})
