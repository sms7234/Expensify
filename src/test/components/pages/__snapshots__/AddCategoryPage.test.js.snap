// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddCategoryPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
         Add Category 
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <CategoryForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
